### ISSUE TITLE
Technology/tool-specific primitives/RTL code for security hardening

### ISSUE THREAD

**Issue by vogelpi** [lowRISC/opentitan#4238]
With #3851, we have added synthesis attributes to our masked AES S-Boxes to make sure Vivado synthesis doesn't aggressively optimize the implementation and thereby accidentally defeat the masking.

These attributes are tool specific and it is not yet clear what is is the best way to support multiple tools and technologies. For example, some synthesis tools allow to translate synthesis attributes (Yosys) while others simply ignore unsupported attributes (DC).

To support multiple tools and technologies, different options have been mentioned so far:
1. Have a separate, synthesis-tool-specific implementation of the security critical module and select the implementation based on a parameter or FuseSoC .core-file. See https://github.com/lowRISC/opentitan/pull/3851#issuecomment-712684769 . This leads to large code duplication and should be avoided.
2. Abstract critical building blocks as tech-specific primitives as we do e.g. with clock gates already, and add the tool/tech-specific attributes inside these primitives, see @msfschaffner 's [comment](https://github.com/lowRISC/opentitan/pull/3851#issuecomment-726942409) .

The idea of this issue is to discuss options and agree on a common path forward.

**Comment by vogelpi** [comment]
This might be of interest for @msfschaffner , @tjaychen , @cdgori , @imphil

**Comment by imphil** [comment]
Can you give an example of how many attributes you typically need? And how you set them? Is this something you do once, confirm after the synthesis has run, and then don't touch again (since they are quite fragile), or do you already know when you read the algorithmic description that you'll need to add certain keep attributes?

To add to the list of options:
3. An external constraint file.

**Comment by imphil** [comment]
So effectively the low-level primitives would give us
1. A place to put tool-specific synthesis attributes that provide a good starting point.
2. A marker/name to use in constraint files, and to look for in various downstream synthesis results/netlists.

Sounds like a good approach to me! (Not that I have strong feelings on any of them.)

**Comment by msfschaffner** [comment]
Sorry a bit late to the party.

This solution is already partially in place, but it can be extended and structured a bit better. For instance, some of the technology primitives such as clock inverters / buffers and flops have a specific name prefix `u_size_only*` that our DC synthesis script looks for when applying the synthesis constraints:

https://github.com/lowRISC/opentitan/blob/545ffb85c0e04ae3a33ac3db507cc3fc3703f892/hw/top_earlgrey/syn/constraints.sdc#L194-L198

We have only used `size_only` attributes so far since we still want the synthesis tool to do automatic up and down sizing during optimization.

To @imphil and @cdgori points above, I think it make sense to:

1) first list out all std cell gates that need to be constrained using attributes. In my view this should encompass all hand instantiated std cells since the only reason to instantiate a cell by hand is actually because you want to make absolutely sure that that cell type is being used and not optimized away. So we could start with this list here:
    - clock_buf (already exists)
    - clock_inv (already exists)
    - clock_mux2 (already exists)
    - flop (already exists)
    - buf (new)
    - inv (new)
    - xor2, xor3 (new)
    - xnor2, xnor3 (new)
    - and2, and3 (new)
    - or2, or3 (new)
    - nor2, nor3 (new)
    - mux2 (new)

2) create generic, xilinx and asic versions thereof, where the generic and Xilinx ones use the "keep" RTL attribute that Vivado and Yosis understand (we may not even need a Xilinx version for all of them if the generic and Xilinx version would be identical). For the ASIC versions, we continue using the `u_size_only_*` naming convention I mentioned in order to be able to easily search for these cells in TCL synthesis and P&R scripts.

3) think about the audit process and how that could be automated. it is possible to report all cells with a given attribute in DC (I would assume Vivado can do that as well?), and we could add some additional post-synthesis check scripting that checks that all important instances are listed. for that, we would have to maintain something like a list of hierarchical patterns that should be matched.

WDYT?

As for the constraints hand-off to P&R, this is a good point that we should bring up with Nuvoton in one of our upcoming meetings (CC @tjaychen).

**Comment by tjaychen** [comment]
the list sounds good to me, but I don't think we necessarily need to construct all of them in one shot now (especially the 3 input ones), and we can do it as the need comes up.  Generally with xors, inverters, buffers and muxes I think we can pretty much do everything we need. 

In terms of the constraints hand-off, I'm not really sure PnR for example needs to be in our current scope.  Since that's not really something we are going to touch (for the time being). So point 3 above to me is going to be pretty important.

Another note is that we might want to consider separating primitive names for cells that just need to be hand-instantiated as an anchor point, vs cells that need to be hand instantiated for security reasons.  It might be possible that other than just existence checks, we may do more with security cells (bury them in metal layers for example), so we probably want a way to differentiate the two. 

Does it make sense to have a vanilla version vs a secure version? Or could we accomplish the same by passing some kind of string via parameters? It feels like having a dedicated version might be cleaner.

**Comment by vogelpi** [comment]
Thanks everybody for your input! The proposed strategy sounds good to me. I agree that we probably don't need to construct all the primitives now, but instead as they become useful.

Regarding the distinction between cells serving as anchor points and for security, this is probably best done by name. We could have a `prim_sec_mux2` and a `prim_mux2`. This may lead so some duplication but I expect the types of cells we need for anchor points to be limited. I don't expect that we will need e.g. an xnor2 or xnor3 for this purpose. But what I think we will need is a parameter to support multiple widths. Otherwise, the code will get super unreadable. For AES, we need e.g. 1,2,4 and 8 bit wide xor2.

Regarding Michael's Point No. 3: Yes, Vivado supports the `get_cells/nets` Tcl command to query objects and also search for properties. However, a quick test showed that the `keep` attribute cannot be looked up. The corresponding nets are there, but the property view doen't list the keep property anymore. So, at least for Vivado, we have to do a name based search.

**Comment by msfschaffner** [comment]
Distinction by name sounds good to me as well!

@vogelpi do you want to go ahead and define the modules that you need? I can go in and add the ASIC counterparts...

**Comment by vogelpi** [comment]
Yes @msfschaffner, I will do that. But probably only while implementing the next masked S-Box in parallel. Before this can happen, I have to do some readings and experiments. If you could support me with the ASIC counterparts, that would be great!

**Comment by msfschaffner** [comment]
Sure, SGTM!

**Comment by vogelpi** [comment]
Yes, we've now a flow for this following proposal No. 2 and we keep adding new primitives as needed. Let's close this.
